iModelCore/ecobjects/src/ExpressionContext.cpp

```diff
@@ -509,6 +503,17 @@ static ExpressionStatus GetInstanceValue (EvaluationResultR evalResult, uint32_t
         evalResult.Clear();
         if (NULL == currentProperty)
             {
+            if (accessString.Equals("ECInstanceId"))
+                {
+                evalResult = ECValue(BeInt64Id::FromString(instance.GetInstanceId().c_str()).ToHexStr().c_str());

```

Comment:
I believe this needs some error handling. BeInt64Id::FromString() takes an optional BentleyStatus pointer as second parameter, and that returns ERROR for invalid IDs. Like, when it's an empty string, or a negative number or decimal number, or alphanumeric letters.
We should cover this scenario with tests. What is supposed to happen with non-numeric Instance IDs? Also, maybe we'll need a non-hex version of this in the future.

Reply:
> I believe this needs some error handling. BeInt64Id::FromString() takes an optional BentleyStatus pointer as second parameter, and that returns ERROR for invalid IDs. Like, when it's an empty string, or a negative number or decimal number, or alphanumeric letters. We should cover this scenario with tests. What is supposed to happen with non-numeric Instance IDs?

Do we have any real scenarios in iTwin.js when an `ECInstance` has empty, negative, decimal, etc. `ECInstanceId`?

> Also, maybe we'll need a non-hex version of this in the future.

We could change result of these expressions to `int64` and then introduce a function symbol `Id64ToHex` or similar to convert it to hex. How does that sound?

Reply:
I'll discuss with the rest of the team in our call and get back to you. We try to keep the old imodel02 and imodel-native code in sync, so even if we don't have non-number Ids in iTwin, this code will probably affect others. Since the API uses strings, it seems very possible we will run into the case.

Reply:
We discussed and it seems safe to assume only numbers here, I'm closing the discussion.

Reply:
I think it's fine to do this.  ids are always ints in the context of an iModel.  Even on imodel02 ids are always ints.  When elements are brought over from dgn side they may have their v8 ids but calculated properties are stripped and not run so this won't possibly impact that.